pkg/controller/statefulset/stateful_pod_control.go

```diff
@@ -135,15 +205,98 @@ func (spc *realStatefulPodControl) UpdateStatefulPod(set *apps.StatefulSet, pod
 	return err
 }
 
-func (spc *realStatefulPodControl) DeleteStatefulPod(set *apps.StatefulSet, pod *v1.Pod) error {
-	err := spc.client.CoreV1().Pods(set.Namespace).Delete(context.TODO(), pod.Name, metav1.DeleteOptions{})
+func (spc *StatefulPodControl) DeleteStatefulPod(set *apps.StatefulSet, pod *v1.Pod) error {
+	err := spc.objectMgr.DeletePod(pod)
 	spc.recordPodEvent("delete", set, pod, err)
 	return err
 }
 
+// ClaimsMatchRetentionPolicy returns false if the PVCs for pod are not consistent with set's PVC deletion policy.
+// An error is returned if something is not consistent. This is expected if the pod is being otherwise updated,
+// but a problem otherwise (see usage of this method in UpdateStatefulPod).
+func (spc *StatefulPodControl) ClaimsMatchRetentionPolicy(set *apps.StatefulSet, pod *v1.Pod) (bool, error) {
+	ordinal := getOrdinal(pod)
+	templates := set.Spec.VolumeClaimTemplates
+	policy := getPersistentVolumeClaimRetentionPolicy(set)
+	// If PVCs are not deleted on scaledown, they are expected to exist.
+	claimShouldBeRetained := policy.WhenScaled == apps.RetainPersistentVolumeClaimRetentionPolicyType
+	for i := range templates {
+		claimName := getPersistentVolumeClaimName(set, &templates[i], ordinal)
+		claim, err := spc.objectMgr.GetClaim(set.Namespace, claimName)
```

Comment:
I didn't catch this... but this is a live read, on every single update loop?  That's concerning and a scalability issue.  I would *not* expect the controller loop to do a live read on each potential statefulset level, but instead deal with the consequences of the stale cache.

I don't think I could tag with this logic in place as is.

Reply:
Ok, objectMgr is reading from caches and writing direct, so this is ok.  Sorry, missed that intentional change.

Reply:
Discussed offline, the objectMgr.GetClaim hits the lister and is not a live read.